File: TopInterview150/Strings/Arrays/ReverseWordsInAString/solution.cpp
class Solution {
public:
    string reverseWords(string s) {
        string word = "";
        string sentence = "";
        for (int i = 0; i < s.size(); i++)
        {
            if (s[i] == ' ' && word == "")
            {
                continue;
            }
            else if (s[i] == ' ')
            {
                sentence = word + " " + sentence;
                word = "";
            }
            else
            {
                word += s[i];
            }
        }
        if (word != "")
        {
            sentence = word + " " + sentence;
        }
        return sentence.substr(0, sentence.size()-1);
    }
};

File: TopInterview150/Strings/Arrays/ReverseWordsInAString/solution.py
class Solution(object):
    def reverseWords(self, s):
        s = s.split(' ')
        result = []
        for x in s:
            result.append(x)
        result.reverse()
        result = ' '.join(result).strip()
        return ' '.join(result.split())


""" works, slower and is more fine grained than what python provides advantages for
class Solution(object):
    def reverseWords(self, s):
        "1""
        :type s: str
        :rtype: str
        "1""
        word = ""
        sentence = ""
        for x in range(len(s)):
            if s[x] == ' ' and word == "":
                continue
            elif s[x] == ' ':
                sentence = word + " " + sentence
                word = ""
            else:
                word += s[x]
        if word != "":
            sentence = word + " " + sentence
        return sentence[:-1]
"""